Gep 2907 update - TLS mode and allowed routes (#3458)
kubernetes-sigs/gateway-api: geps/gep-2907/index.md

The above diagram depicts these four segments as edges in a graph.

### TLS mode

TLS can be configured with two distinct modes:

* **Terminate**: the TLS connection is instantiated between the frontend and the
  Gateway. The connection between the Gateway and the backend is left unencrypted
 unless a new TLS connection between the two entities is configured via BackendTLSPolicy.
* **Passthrough**: the TLS connection is instantiated between the frontend and the
  backend. The traffic flows through the Gateway encrypted, and the Gateway is not

>   able to decrypt or inspect the TLS stream.

Reviewer: ```suggestion
  able to decrypt or inspect the encrypted portions of the TLS stream.
```

May be a bit more precise? Unless the ClientHello isn't considered a part of the stream

Reviewer 2: To what extent might SNI routing be complicated by encrypted client hello? It looks like "Split Mode" might still be designed to work with passthrough mode for the rest of the stream? https://www.ietf.org/archive/id/draft-ietf-tls-esni-17.html

Author: I started to look down the rabbit hole and it looks quite a complex (and not yet completely defined) topic. The TL;DR is that it's complex problem, as the client-facing server does never have access to the SNI (this is the whole point of ECH), so it needs to use alternative ways to figure out where to send traffic (a lot of details in the RFC). 

Even if definitely interesting, I think this is way beyond the purpose of this update. Accepting the suggestion from @keithmattix though, as it's definitely more precise.

### Context

The `*Route` objects refer the Gateway (or a specific Listener of the Gateway) they
want to be attached to. A successful attachment can be granted by one of the two
following conditions:

* the `Listener.AllowedRoutes` field allows that specific Route `GroupKind` to be
  attached to it, or
* the `Listener.Protocol` field allows that specific Route to be attached to the
  Listener. This applies only in case the Gateway's AllowedRoutes field is not set.

In case the First condition is not satisfied (the Route references a Gateway that
does not explicitly allow such an attachment), the Route `Accepted` condition
is set to False with reason `NotAllowedByListeners`. In case the Route references
a Gateway with no Listener matching the protocol needed by the Route, the Route
`ResolvedRefs` condition is set to False with reason `UnsupportedProtocol`.

### What Routes are allowed by Listeners with TLS

The following is a summary of all the Routes and the TLS termination mode they support.

| Route | TLS with Terminate | TLS with Passthrough |
|-|-|-|
| `HTTPRoute` | yes | no |
| `GRPCRoute` | yes | no |
| `TLSRoute` | yes | yes |

> | `TCPRoute` | no | no |

Reviewer 3: Why not TLS with passthrough here?

Reviewer 4: If you are doing TLS passthrough and a TCPRoute, its no different than just TCP + TCPRoute. Unless you want to say we allow TLSRoute and TCPRoute to bind to the same listener here, then you would have something like

1. Highest priority: TLS route matching SNI
2. Lowest priority: the (at most 1) TCP route
?

| `UDPRoute` | no | no |

> [!NOTE]
> When the traffic is routed to the backend via `TCPRoute`, the packets
> are left untouched by the gateway. In order to terminate the TLS connection to
> the gateway and forward the traffic unencrypted to the backend, a `TLSRoute` configured

> > with `Terminate` as TLS mode has to be used.

Reviewer 3: This is a bit confusing to me. Why does the route type matter here? In the case of TLS passthrough, it seems like you'd have two options if you wanted to use SNI to select the backend:

1. Multiple Listeners with different hostnames specified
2. One Listener with multiple TLSRoutes attached, each specifying a different SNI

On the other hand, if you wanted to do TLS termination, you'd have the following options:

1. Route based on HTTP attributes: HTTPRoute
2. Send all traffic attached to a Listener to the same set of backends: HTTPRoute or TCPRoute

Writing that all out makes me continue to think that our <L7 route types may not be entirely necessary (x-ref the [related doc](https://docs.google.com/document/d/1qj7Xog2t2fWRuzOeTsWkabUaVeOF7_2t_7appe8EXwA/edit?tab=t.0#heading=h.w311n4l5qmwk)).

Reviewer 4: There is a third plausible option, you could do "One Listener with multiple TLSRoutes attached, each specifying a different SNI" but still use terminate. I mean hypothetically, I believe we say this isn't the semantics of the routes.

Reviewer 3: @howardjohn that's a good point, it's certainly possible with the API surface we currently have, do you know of any common use cases for that?

Author: This is good discussion, I have a question: In the case of encrypted L4 traffic, why users should use `TCPRoute` with TLS Terminate over`TLSRoute`? This is one of the main aspects I'm trying to capture with this GEP update. The assumption I'm making is: in case you need TLS, use either `HTTPRoute` or `TLSRoute`, depending on your needs and use-case.

Reviewer: So why do we need both TCPRoute and TLSRoute? Could TCPRoute just have optional TLS config? It's a bit of a sidebar, but the omission. Of TCPRoute here prompts the question I think

Reviewer 4: > @howardjohn that's a good point, it's certainly possible with the API surface we currently have, do you know of any common use cases for that?

So we don't support the equivilent of this in Istio so I don't have as much concrete data on this. I can only hypothesize that you might want to have 1 GW with a broad wildcard and then scoped TLSRoutes owned by a different persona.

Basically the same story as with HTTP

Reviewer 5: > > @howardjohn that's a good point, it's certainly possible with the API surface we currently have, do you know of any common use cases for that?
> 
> So we don't support the equivilent of this in Istio so I don't have as much concrete data on this. I can only hypothesize that you might want to have 1 GW with a broad wildcard and then scoped TLSRoutes owned by a different persona.
> 
> Basically the same story as with HTTP

Yes, we have that use case:
- One GW with wildcard
- TLS termination for HTTP configured in the GW so devs can just add HTTPRoutes to their app deployments
- The app also comes with a MQTT broker. That connection can be terminated the same way in the GW, but at the moment there is no way for the devs to just add a TCPRoute to their several apps. Since it's lacking a discriminator.

For now we need to create separate GW with exact hostnames for each MQTT broker next to the wildcard GW. And for each we need to extend the ReferenceGrant to the wildcard certificate. A less complicated way is much appreciated :)

Either solution is fine for us:
- TLSRoute allowed for terminated TLS connections
- or a TCPRoute with optional hostname field

Author: Trying to wrap up this whole discussion here. I think we may need to scope down which route kinds are allowed for each protocol first, because this is something we never (as far as I am concerned) well-defined. What I see as the simplest and most obvious choice here is:

| Protocol  |  Routes | TLS Terminate | TLS Passthough |
|---|---|---|---|
| HTTP  | HTTPRoute  |  yes | no |
|  GRPC | GRPCRoute  | yes | no |
|  TLS | TLSRoute  | yes | yes | 
|  TCP | TCPRoute  | no | no |
|  UDP | UDPRoute  | no | no |

We could definitely discuss if some more complex combinations may exist, such as TCPRoute with protocol TLS and TLS passthrough allowed (described in a comment above), but unless clear demands and use-cases are defined and asked, I'd stick with the most linear and simplest one for now.

This is something we can iterate on and improve over time if new use-cases will pop up.

Author: In case this proposal for now looks ok to y'all, I'll use this table in the GEP.

Reviewer 4: I really don't feel like this is the right way to go. It becomes an inconsistent model.

The way I view things, is there are 3 phases of a gateway.

* Downstream accepting -- accepting things from the clinet
* Internal routing and process
* Upstream sending -- what we send to the client

In many cases, these can all be the same protocol. For example, we may accept HTTP, process HTP, and then send HTTP. Easy.

If we look at TLS termination, though, its really more like this:
* Downstream: accept TLS, terminate it, "unwrapping" the inner protocol
* Internal processing: routing based on the **inner protocol**
* Upstream: either forward the inner protocol, or with BackendTLS, possible wrap in TLS

So for HTTPS with terminate:
* Downstream: accepts HTTPS, unwraps to HTTP
* Routing processes HTTP, as HTTPRoute
* Upstream: send as HTTP

So for TLS with terminate, if we are following this model:
* Downstream: accepts TLS, unwrap to TCP
* Routing: process opaque TCP as TCPRoute
* Upstream: send as opaque TCP data

Reviewer 4: I am not entirely opposed to allowing TLSRoute on terminate - it has value for the exact same reasons ListenerSet does (though perhaps that makes them redundant) - but TCPRoute ought to be allowed on TLS termination.

Another way to look at it: as an application developer, my app is exposed as <HTTP/TCP> so I write a <HTTP/TCP>Route. If the gateway admin decides to add TLS security to the gateway, thats not (necessarily) my problem and I shouldn't have to change my routes.

Reviewer 3: Just catching up on this - @mlavacca's updated table in the previous comment LGTM, thanks for all the great discussion on this thread!

Author: yep, addressed in 358b483783b56330988272129d88507bd33acb4d. Closing the thread

Reviewer 3: Thanks @mlavacca! I ended up looking at the PR, forgetting the context and having to reread this thread to understand how we got to this conclusion. Do you mind adding some of this context/rationale in the PR?

Author: Yep, I've added some context to the PR description.

> > with `Terminate` as TLS mode has to be used.

Reviewer 2: ```suggestion
> with `Terminate` as TLS mode must be used.
```

want to be attached to. A successful attachment can be granted by one of the two
following conditions:

* the `Listener.AllowedRoutes` field allows that specific Route `GroupKind` to be
  attached to it, or
* the `Listener.Protocol` field allows that specific Route to be attached to the
  Listener. This applies only in case the Gateway's AllowedRoutes field is not set.

In case the First condition is not satisfied (the Route references a Gateway that
does not explicitly allow such an attachment), the Route `Accepted` condition
is set to False with reason `NotAllowedByListeners`. In case the Route references
a Gateway with no Listener matching the protocol needed by the Route, the Route
`ResolvedRefs` condition is set to False with reason `UnsupportedProtocol`.

### What Routes are allowed by Listeners with TLS

The following is a summary of all the Routes and the TLS termination mode they support, along
with the compatible protocol.

| Protocol  |  Routes | TLS Terminate | TLS Passthough |
|-----------|---------|---------------|----------------|
| HTTP  | `HTTPRoute` | no  | no  |
| HTTPS | `HTTPRoute` | yes | no  |

> | GRPC  | `GRPCRoute` | yes | no  |

Reviewer 4: sorry I missed this until now - grpc is not a protocol in the API  https://gateway-api.sigs.k8s.io/reference/spec/#gateway.networking.k8s.io/v1.ProtocolType

This should just be under HTTP/HTTPS I think with the same rules as HTTPRoute

Author: good catch, her is the fix: 8ab1bc21b00fc50c9bfbe22b4a6f85531449b9b1
